Objects/dictobject.c

```diff
@@ -1496,6 +1496,97 @@ _Py_dict_lookup_threadsafe(PyDictObject *mp, PyObject *key, Py_hash_t hash, PyOb
     return ix;
 }
 
+Py_ssize_t
+_Py_dict_lookup_threadsafe_stackref(PyDictObject *mp, PyObject *key, Py_hash_t hash, _PyStackRef *value_addr)
+{
+    PyDictKeysObject *dk;
+    DictKeysKind kind;
+    Py_ssize_t ix;
+
+    ensure_shared_on_read(mp);
+
+    dk = _Py_atomic_load_ptr(&mp->ma_keys);
+    kind = dk->dk_kind;
+
+    if (kind != DICT_KEYS_GENERAL) {
+        if (PyUnicode_CheckExact(key)) {
+            ix = unicodekeys_lookup_unicode_threadsafe(dk, key, hash);
+        }
+        else {
+            ix = unicodekeys_lookup_generic_threadsafe(mp, dk, key, hash);
+        }
+        if (ix == DKIX_KEY_CHANGED) {
+            goto read_failed;
+        }
+
+        if (ix >= 0) {
+            if (kind == DICT_KEYS_SPLIT) {
+                PyDictValues *values = _Py_atomic_load_ptr(&mp->ma_values);
+                if (values == NULL)
+                    goto read_failed;
+
+                uint8_t capacity = _Py_atomic_load_uint8_relaxed(&values->capacity);
+                if (ix >= (Py_ssize_t)capacity)
+                    goto read_failed;
+
+                *value_addr = PyStackRef_FromPyObjectNew(values->values[ix]);
+                if (PyStackRef_IsNull(*value_addr)) {
+                    goto read_failed;
+                }
+
+                if (values != _Py_atomic_load_ptr(&mp->ma_values)) {
+                    goto read_failed;
+                }
+            }
+            else {
+                *value_addr = PyStackRef_FromPyObjectNew(DK_UNICODE_ENTRIES(dk)[ix].me_value);
+                if (PyStackRef_IsNull(*value_addr)) {
+                    goto read_failed;
+                }
+
+                if (dk != _Py_atomic_load_ptr(&mp->ma_keys)) {
+                    goto read_failed;
+                }
+            }
+        }
+        else {
+            *value_addr = PyStackRef_NULL;
+        }
+    }
+    else {
+        ix = dictkeys_generic_lookup_threadsafe(mp, dk, key, hash);
+        if (ix == DKIX_KEY_CHANGED) {
+            goto read_failed;
+        }
+        if (ix >= 0) {
+            *value_addr = PyStackRef_FromPyObjectNew(DK_ENTRIES(dk)[ix].me_value);
+            if (PyStackRef_IsNull(*value_addr)) {
+                goto read_failed;
+            }
+            if (dk != _Py_atomic_load_ptr(&mp->ma_keys)) {
+                goto read_failed;
+            }
+        }
+        else {
+            *value_addr = PyStackRef_NULL;
+        }
+    }
+
+    return ix;
+
+    PyObject *value;
```

Comment:
I assume that it's because you cannot have the declaration just after a label right? However, you should be able to put the declaration after the label by having something like:

```C
read_failed:; /* semicolon to allow a declaration after a label */ 
	PyObject *value;
```

Reply:
Let's not do that, it looks weird :).